lightning/src/routing/scoring.rs

```diff
@@ -460,25 +466,420 @@ impl<T: Time> Readable for ChannelFailure<T> {
+	/// liquidity knowledge will never decay except when the bounds cross.
+	pub liquidity_offset_half_life: Duration,
+}
+
+impl_writeable_tlv_based!(ProbabilisticScoringParameters, {
+	(0, liquidity_penalty_multiplier_msat, required),
+	(2, liquidity_offset_half_life, required),
+});
+
+/// Accounting for channel liquidity balance uncertainty.
+///
+/// Direction is defined in terms of [`NodeId`] partial ordering, where the source node is the
+/// first node in the ordering of the channel's counterparties. Thus, swapping the two liquidity
+/// offset fields gives the opposite direction.
+struct ChannelLiquidity<T: Time> {
+	/// Lower channel liquidity bound in terms of an offset from zero.
+	min_liquidity_offset_msat: u64,
+
+	/// Upper channel liquidity bound in terms of an offset from the effective capacity.
+	max_liquidity_offset_msat: u64,
+
+	/// Time when the liquidity bounds were last modified.
+	last_updated: T,
+}
+
+/// A snapshot of [`ChannelLiquidity`] in one direction assuming a certain channel capacity and
+/// decayed with a given half life.
+struct DirectedChannelLiquidity<L: Deref<Target = u64>, T: Time, U: Deref<Target = T>> {
+	min_liquidity_offset_msat: L,
+	max_liquidity_offset_msat: L,
+	capacity_msat: u64,
+	last_updated: U,
+	now: T,
+	half_life: Duration,
+}
+
+impl<G: Deref<Target = NetworkGraph>, T: Time> ProbabilisticScorerUsingTime<G, T> {
+	/// Creates a new scorer using the given scoring parameters for sending payments from a node
+	/// through a network graph.
+	pub fn new(params: ProbabilisticScoringParameters, network_graph: G) -> Self {
+		Self {
+			params,
+			network_graph,
+			channel_liquidities: HashMap::new(),
+		}
+	}
+
+	#[cfg(test)]
+	fn with_channel(mut self, short_channel_id: u64, liquidity: ChannelLiquidity<T>) -> Self {
+		assert!(self.channel_liquidities.insert(short_channel_id, liquidity).is_none());
+		self
+	}
+}
+
+impl Default for ProbabilisticScoringParameters {
+	fn default() -> Self {
+		Self {
+			liquidity_penalty_multiplier_msat: 1000,
+			liquidity_offset_half_life: Duration::from_secs(3600),
+		}
+	}
+}
+
+impl<T: Time> ChannelLiquidity<T> {
+	#[inline]
+	fn new() -> Self {
+		Self {
+			min_liquidity_offset_msat: 0,
+			max_liquidity_offset_msat: 0,
+			last_updated: T::now(),
+		}
+	}
+
+	/// Returns a view of the channel liquidity directed from `source` to `target` assuming
+	/// `capacity_msat`.
+	fn as_directed(
+		&self, source: &NodeId, target: &NodeId, capacity_msat: u64, half_life: Duration
+	) -> DirectedChannelLiquidity<&u64, T, &T> {
+		let (min_liquidity_offset_msat, max_liquidity_offset_msat) = if source < target {
+			(&self.min_liquidity_offset_msat, &self.max_liquidity_offset_msat)
+		} else {
+			(&self.max_liquidity_offset_msat, &self.min_liquidity_offset_msat)
+		};
+
+		DirectedChannelLiquidity {
+			min_liquidity_offset_msat,
+			max_liquidity_offset_msat,
+			capacity_msat,
+			last_updated: &self.last_updated,
+			now: T::now(),
```

Comment:
Wow, that's a lot of system-time-gets batman! I guess its pretty quick cause it rtdsc's on x86 usually when its warm, but we should probably fix this later. Maybe add a comment here noting that when we remove old entries, we should just decay them then and not bother decaying during lookups.

Reply:
> Wow, that's a lot of system-time-gets batman! I guess its pretty quick cause it rtdsc's on x86 usually when its warm, but we should probably fix this later.

This should be the similar behavior as `Scorer::decayed_penalty_msat`, right?

> Maybe add a comment here noting that when we remove old entries, we should just decay them then and not bother decaying during lookups.

Not sure I follow. The penalty may not be accurate if we don't decay during lookup. Could you clarify what you mean by "remove old entries"?

Reply:
> This should be the similar behavior as Scorer::decayed_penalty_msat, right?

I think so? Doesn't mean its great, though. IIRC it can be quite cheap on x86, but on other platforms it can be a full syscall.

> Not sure I follow. The penalty may not be accurate if we don't decay during lookup. Could you clarify what you mean by "remove old entries"?

I'm assuming when we implement the "remove entries for closed channels" logic, we'll instead just remove entries on a timer that have decayed almost all the way to zero. Once we have a timer, we can just do the decays on the timer, instead of doing the decays while routing.

Reply:
Gotcha. Yeah, that would be a simple way to remove closed channels. I'd hope there would be a way to use the scorer without a background thread, though.

Another thought would be to push `Time` into either the `Score` or `find_route` interface so `T::now` could be called once and passed in as needed. Or even build it in to `LockableScore` in some way such that locking gives back some `(S, T)` implementing `Score`.

Reply:
Yea, that would work too. I dunno what the right answer is. We already require a background thread for the network graph to expire channels, which we'd discussed at the time either doing there or on serialization and decided background thread was simpler. I guess we could have an explicit "channel is dead" call here like we do for the network graph, but that feels like about as much work as just hooking it up to the background thread/a timer? Anyway, its something we can discuss another time, I don't think this needs fixing here, but I worry this is only practically performant on x86.